### DISCUSSION TITLE
Idea for new Aliases cops

### DISCUSSION THREAD

**Discussion by noraj** [rubocop/rubocop#13082] [Ideas]
I would like to suggest an idea 💡 : having more aliases style cops!

There are many aliases in Ruby

- `Hash#select` ↔️ `Hash#filter`
- `Hash#each_pair` ↔️ `Hash#each`
- `Hash#include?` ↔️ `Hash#has_key?`
- `Hash#value?` ↔️ `Hash#has_value?`
- `Array#push` ↔️ `Array#append`
- `Array#collect` ↔️ `Array#map`
- `Array#find_index` ↔️ `Array#index`
- etc. just to name a few

I looked for existing alias cops, and it seems there are no alias cops for Ruby and 2 alias cops for RoR [Rails/ActiveRecordAliases](https://docs.rubocop.org/rubocop-rails/cops_rails.html#railsactiverecordaliases) and [Rails/ActiveSupportAliases](https://docs.rubocop.org/rubocop-rails/cops_rails.html#railsactivesupportaliases).

Note: [Style/Alias](https://docs.rubocop.org/rubocop/cops_style.html#stylealias) is for enforcing the use of either `#alias` or `#alias_method` not to enforce the style of aliases usage.

It would be nice to have one or several aliases style cops for the ones who want to implement a consistent naming convention.

But then, a new question arises: are we going to make a group cop with all aliases (maybe too large), one cop per group / class / module (how to split that?) or one cop per alias (probably not, that would be too much)?

- **Option 1**: `Style/Aliases` for all Ruby standard library aliases (same with `RSpec/Aliases`, `Minitest/Aliases`)
- **Option 2**: `Style/ArrayAliases`, `Style/HashAliases`, etc. aliases grouped by class / module
- **Option 3**: `Style/ArrayPushAlias`, `Style/ArrayCollectAlias`, etc. one cop per alias

Then I see one parameter so that it would be configurable:

`EnforceAliases`:

- `never` (default): using aliases is forbidden, must use only original method name
- `always`: using original method name is forbidden, must use only aliases
- `mixed`: allow using both aliases and original method names

The depending on the grouping chosen:

- if **Option 1** then having an `AllowedClasses` parameter to disable the cop for a specific class `AllowedClasses: [Hash]`.
- if **Option 2** then having an `AllowedMethods` parameter to disable the cop for a specific method `AllowedMethods: [push, collect]`.
- if **Option 3** then I see no additional parameter needed

**Comment by Earlopain** [comment]
RuboCop will have trouble with types here I think, bunch of false positives. It's lacking the type information for this to properly work.

I guess it may make sense to have a generic cop that checks for `methodA` -> `methodB` but enforcing one over the other by default seems bad. I would rather have it 100% configured by the user with no defaults at all. Basically:

```yml
Style/AliasedMethods:
  Enabled: true
  Aliases:
    find_index: index
    each_pair: pair
    etc...
```

**Reply by noraj** [reply]
Yeah, why not no default.

**Comment by dvandersluis** [comment]
I don't think any new cop is needed here.

You can add your own methods to `Style/CollectionMethods`, and some of these are already covered.

```yaml
Style/CollectionMethods:
  PreferredMethods:
    filter: 'select'
```

The Hash methods (`#has_key?` vs `key?` and `#has_value?` vs `value?`) are covered by `Style/PreferredHashMethods`

See https://docs.rubocop.org/rubocop/cops_style.html#stylecollectionmethods and https://docs.rubocop.org/rubocop/cops_style.html#stylepreferredhashmethods
